### PR TITLE
Support custom TLS certificates for old Puppetserver

### PR DESCRIPTION
**Pull request by bastelfreak**

### CODE REVIEW

File: lib/puppet/catalog-diff/compilecatalog.rb [hunk 1]
```diff
@@ -97,14 +97,14 @@ def compile_catalog(node_name, server, certless)
       end
 
       uri = URI("https://#{server}:#{port}#{endpoint}")
-      Puppet.debug("Connecting to server: #{uri}")
```
Comment (bastelfreak) [inline lib/puppet/catalog-diff/compilecatalog.rb:100]: I don't think that this debug statement is required. The http client itself will log this at debug level already.

File: lib/puppet/catalog-diff/compilecatalog.rb [hunk 2]
```diff
@@ -97,14 +97,14 @@ def compile_catalog(node_name, server, certless)
...
       begin
         ret = if certless
-                Puppet.runtime[:http].post(uri, body.to_json, headers: headers)
+                Puppet.runtime[:http].post(uri, body.to_json, headers: headers, options: { ssl_context: ssl_context })
               else
-                Puppet.runtime[:http].get(uri, headers: headers)
+                Puppet.runtime[:http].get(uri, headers: headers, options: { ssl_context: ssl_context })
               end
-        raise "HTTP request to PuppetDB failed with: HTTP #{ret.code} - #{ret.reason}" unless ret.success?
+        raise "HTTP request to Puppetserver #{server} failed with: HTTP #{ret.code} - #{ret.reason}" unless ret.success?
```
Comment (bastelfreak) [inline lib/puppet/catalog-diff/compilecatalog.rb:107]: this was a typo previously. the request was always going to the puppetserver, never puppetdb

File: lib/puppet/catalog-diff/compilecatalog.rb [hunk 3]
```diff
@@ -97,14 +97,14 @@ def compile_catalog(node_name, server, certless)
       end
 
       uri = URI("https://#{server}:#{port}#{endpoint}")
-      Puppet.debug("Connecting to server: #{uri}")
+      ssl_context = Puppet::CatalogDiff::Tlsfactory.ssl_context(tls_cert, tls_key, tls_ca)
       begin
         ret = if certless
-                Puppet.runtime[:http].post(uri, body.to_json, headers: headers)
+                Puppet.runtime[:http].post(uri, body.to_json, headers: headers, options: { ssl_context: ssl_context })
```
Comment (evgeni) [inline lib/puppet/catalog-diff/compilecatalog.rb:103]: doesn't this now make `certless` require certs (to be provided)?

Comment (bastelfreak) [reply lib/puppet/catalog-diff/compilecatalog.rb:103]: the name `certless` is a bit bogus here. if this boolean toggle is true, we talk to a puppetserver API with the name `certless`. This endpoint requires a TLS client cert to authenticate. Naming things is hard... 
It's this endpoint: https://puppet.com/docs/puppet/7/server/puppet-api/v4/catalog.html

It allows you to request a catalog for a common name that's not listed in the client cert you authenticate with.

Comment (evgeni) [reply lib/puppet/catalog-diff/compilecatalog.rb:103]: :bulb: 

so really `these-certs-are-made-for-requesting-everything`, not `certless`, gotcha

Comment (bastelfreak) [reply lib/puppet/catalog-diff/compilecatalog.rb:103]: yes :D

**Review by sebastianrakel — APPROVED** [review]

**Review by evgeni — APPROVED** [review]
(I thought I already had this approved, sorry)

**Pull request merged**